compiler/bindings/python/iree/compiler/tools/onnx.py

```diff
@@ -0,0 +1,165 @@
+import os
+import tempfile
+import logging
+import onnx
+
+from dataclasses import dataclass
+from typing import IO, Any, Optional, Union
+from pathlib import Path
+
+from .. import CompilerOptions, InputType, TempFileSaver
+from ..extras import onnx_importer
+from ..ir import Context, StringAttr
+from .binaries import invoke_pipeline
+from .core import build_compile_command_line
+from ...runtime import Config, VmInstance, VmModule, load_vm_module
+
+__all__ = [
+    "compile_saved_model",
+    "compile_model"
+]
+
+
+logger = logging.getLogger(__name__)
+
+
+def load_onnx_model(raw_model: onnx.ModelProto, 
+                    data_prop: bool = True, 
+                    data_dir : IO[bytes] | str | os.PathLike | None = None):
+    # Load the model, with possible external data coming from the default
+    # location, or the location specified on the command line.
+    if data_dir:
+        onnx.load_external_data_for_model(raw_model, str(data_dir))
+
+    # Do shape inference two ways.  First, attempt in-memory to avoid redundant
+    # loading and the need for writing a temporary file somewhere.  If that
+    # fails, typically because of the 2 GB protobuf size limit, try again via
+    # files.  See
+    # https://onnx.ai/onnx/repo-docs/PythonAPIOverview.html#shape-inference-a-large-onnx-model-2gb
+    # for details about the file-based technique.
+
+    # Run the checker to test whether the file is above the threshold for
+    # in-memory shape inference.  If not, go ahead and do the shape inference.
+    try:
+        onnx.checker.check_model(raw_model)
+        inferred_model = onnx.shape_inference.infer_shapes(
+            raw_model, data_prop=data_prop
+        )
+        return inferred_model
+    except ValueError:
+        pass
+
+    # Model is too big for in-memory inference: do file-based shape inference
+    # to a temp file.
+    # Make a temp dir for all the temp files we'll be generating as a side
+    # effect of infering shapes. For now, the only file is a new .onnx holding
+    # the revised model with shapes.
+    with tempfile.TemporaryDirectory() as temp_dir_name:
+        temp_dir_path = Path(temp_dir_name)
+        temp_input_file = temp_dir_path + "/" + "temp.onnx"
+        temp_inferred_file = temp_dir_path + "/" + "temp-inferred.onnx"
```

Comment:
Not to get too into the small details in this initial code review, but when using `Path` from https://docs.python.org/3/library/pathlib.html, you can (and should) just use the `/` operator to join path components.

```suggestion
        temp_dir_path = Path(temp_dir_name)
        temp_input_file = temp_dir_path / "temp.onnx"
        temp_inferred_file = temp_dir_path / "temp-inferred.onnx"
```